In [1]:
# experimental: you can run this cell to see your code full-width
from IPython.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

In [2]:
# cell for pair-programming magic link

#  run this cell without changing it. It only works on Noteable.
! printf "\n\nLink below can be send to someone to share your Noteable session:\n\nhttps://noteable.edina.ac.uk/user/$(jupyter notebook list | grep -oP '(?<=user\/).*(?=\/\?)' )/tree?token=$( jupyter notebook list | grep -oP '(?<=token=).*(?= ::)' )\n\n"



Link below can be send to someone to share your Noteable session:

https://noteable.edina.ac.uk/user/d7i6m6qmi3q5tglmiqr86m/tree?token=OyaJoCPcRPevE70HdUwVrg



# Who survived on Titanic?

In this lab we will use a famous leanring dataset about passangers of titanic. Titanic was a big passager boat which sunk on 15 April 1912. Lots of people died, but we have quite detailed information about passangers and whether they survived. On this lab we will investigate relationship between those variables (eg. wealth, age, gender, cabin number) and survival.

In [2]:
import pandas as pd
import numpy as np

In [3]:
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
# this version of titanic dataset is taken from Kaggle
# https://www.kaggle.com/competitions/titanic/data?select=test.csv
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [4]:
# Getting the unique values of the 'SibSp' column
set(passengers['SibSp'])

{0, 1, 2, 3, 4, 5, 8}

In [7]:
# Getting the unique values of the 'Parch' column
set(passengers['Parch'])

{0, 1, 2, 3, 4, 5, 6}

In [6]:
# Getting the unique values of the 'Embarked' column
set(passengers['Embarked'])

{'C', 'Q', 'S', nan}

In [102]:
# Loading the dataset again, but this time using the 'convert_dtypes' method to 
# convert the columns to the best possible data type
passengers = pd.read_csv('./data/titanic_kaggle_train.csv').convert_dtypes()
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  <NA>      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171     7.25  <NA>        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282    7.925  <NA>        S  
3        0            113803     53.1  C123        S  
4        0            373450     8.05  <NA>        S  
..     ...               ...      ...   ...      ...  
886      0            211536     13.0  <NA>        S  
887      0            112053     30.0   B42        S  
888      2        W./C. 6607    23.45  <NA>        S  
889      0            111369     30.0  C148        C  
890      0            370376     7.75  <NA>        Q  

[891 rows x 12 columns]

# Task 1: Descriptive statistics

Describe the dataset in a set of bulletpoints (use Pandas to give you the answers). Try to backup each bulletpoint with separate pieces of python code.

Below are just some examples to get you started, but explore numbers and types of things (rows, columns). What are max min and mean of the numeric ones? Which columns have missing values and how many? etc.

For exaplanations of what columns mean see https://www.kaggle.com/competitions/titanic/data?select=test.csv

### A. What columns and of what type did you find? How many rows and columns?

In [11]:
# Checking the dimensions of the dataset
passengers.shape

(891, 12)

In [12]:
# Checking the columns of the dataset
passengers.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [13]:
# Checking the data type of the columns output
type(passengers.columns)

pandas.core.indexes.base.Index

In [15]:
# Checking the data type of the dataset
type(passengers)

pandas.core.frame.DataFrame

In [16]:
# Checking information about the dataset
passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    Int64  
 1   Survived     891 non-null    Int64  
 2   Pclass       891 non-null    Int64  
 3   Name         891 non-null    string 
 4   Sex          891 non-null    string 
 5   Age          714 non-null    Float64
 6   SibSp        891 non-null    Int64  
 7   Parch        891 non-null    Int64  
 8   Ticket       891 non-null    string 
 9   Fare         891 non-null    Float64
 10  Cabin        204 non-null    string 
 11  Embarked     889 non-null    string 
dtypes: Float64(2), Int64(5), string(5)
memory usage: 89.8 KB


In [103]:
# Changing the data type of the 'Survived' column to boolean
passengers['Survived'] = passengers['Survived'].astype('bool')

# Changing the data types of columns to category
passengers['Pclass'] = passengers['Pclass'].astype('category')
passengers['Sex'] = passengers['Sex'].astype('category')
passengers['SibSp'] = passengers['SibSp'].astype('category')
passengers['Parch'] = passengers['Parch'].astype('category')
passengers['Embarked'] = passengers['Embarked'].astype('category')

# Renaming the categories of the 'Embarked' column
passengers['Embarked'] = passengers['Embarked'].cat.rename_categories({'C': 'Cherbourg', 'Q': 'Queenstown', 'S': 'Southampton'})
print('Embarked:', passengers['Embarked'].cat.categories)
print('Embarked missing:', passengers['Embarked'].isna().any())

# Renmaing the categories of the 'Pclass' column
passengers['Pclass'] = passengers['Pclass'].cat.rename_categories({1: '1st', 2: '2nd', 3: '3rd'})
print('Pclass:', passengers['Pclass'].cat.categories)

# Checking the information about the dataset
print(passengers.info())
passengers

Embarked: Index(['Cherbourg', 'Queenstown', 'Southampton'], dtype='object')
Embarked missing: True
Pclass: Index(['1st', '2nd', '3rd'], dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    Int64   
 1   Survived     891 non-null    bool    
 2   Pclass       891 non-null    category
 3   Name         891 non-null    string  
 4   Sex          891 non-null    category
 5   Age          714 non-null    Float64 
 6   SibSp        891 non-null    category
 7   Parch        891 non-null    category
 8   Ticket       891 non-null    string  
 9   Fare         891 non-null    Float64 
 10  Cabin        204 non-null    string  
 11  Embarked     889 non-null    category
dtypes: Float64(2), Int64(1), bool(1), category(5), string(3)
memory usage: 50.8 KB
None


PassengerId  Survived Pclass  \
0              1     False    3rd   
1              2      True    1st   
2              3      True    3rd   
3              4      True    1st   
4              5     False    3rd   
..           ...       ...    ...   
886          887     False    2nd   
887          888      True    1st   
888          889     False    3rd   
889          890      True    1st   
890          891     False    3rd   

                                                  Name     Sex   Age SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0     1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0     1   
2                               Heikkinen, Miss. Laina  female  26.0     0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0     1   
4                             Allen, Mr. William Henry    male  35.0     0   
..                                                 ...     ...   ...   ...   
886                              Montvila, Rev. Juozas    male  27.0     0   
887                       Graham, Miss. Margaret Edith  female  19.0     0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  <NA>     1   
889                              Behr, Mr. Karl Howell    male  26.0     0   
890                                Dooley, Mr. Patrick    male  32.0     0   

    Parch            Ticket     Fare Cabin     Embarked  
0       0         A/5 21171     7.25  <NA>  Southampton  
1       0          PC 17599  71.2833   C85    Cherbourg  
2       0  STON/O2. 3101282    7.925  <NA>  Southampton  
3       0            113803     53.1  C123  Southampton  
4       0            373450     8.05  <NA>  Southampton  
..    ...               ...      ...   ...          ...  
886     0            211536     13.0  <NA>  Southampton  
887     0            112053     30.0   B42  Southampton  
888     2        W./C. 6607    23.45  <NA>  Southampton  
889     0            111369     30.0  C148    Cherbourg  
890     0            370376     7.75  <NA>   Queenstown  

[891 rows x 12 columns]

In [21]:
# Printing the mean of the 'Age' column
print("Mean:", passengers.Age.mean())

# Printing the median of the 'Age' column
print("Median:", int(passengers['Age'].median()))

Mean: 29.69911764705882
Median: 28


**Answer A:**

there are 12 columns (+ index) and 891 rows. 

Columns have names: 'PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'

Name and Sex columns are an object (likely a string), so we converted them to `strings`. Age and Fare are Floats, and Sibling/Parent rows are integers.

Age had a mean 29 years, and max and min are ... [you continue]

# Task 2. Answers basic questions: Select some rows, some columns and run some simple statistics

What makes a good **'Basic Question'**? Question which you can explore by selecting only some rows and columns.

Example Basic Questions:

- just select name and age components of the data frame. Sort them by Age and get the names of 10 oldest passangers
- Pick the fares of pasangers who survived and did not survive and compare their ticket class (Pclass), Fare cost and sex. Is there a large difference in the means?

**Itentify 5 Basic Questions** like above and answer them with a simple sentence. Provide python code which will find you that answer. 

Call them 'A', 'B', 'C', ...

### To get you started, here is some starting code. Read it, make sure you understand it. Feel free to modify it.

How to select only some rows and some columns?

Note about referencing: iloc[rows, columns] is used when referencing with numbers, and loc[rows, columns] is used when referencing by names. 

Also colon `:` means `'all of them'`

In [22]:
# first passanger, their data in column indexed 4
passengers.iloc[0,4]

'male'

In [23]:
passengers.loc[:,'Survived'].value_counts() # Looking for survive status

Survived
False    549
True     342
Name: count, dtype: int64

In [29]:
passengers.iloc[0:5,[2,3,-1]] # first 5 passangers, only columns 3, 4 and the last one

Pclass                                               Name     Embarked
0    3rd                            Braund, Mr. Owen Harris  Southampton
1    1st  Cumings, Mrs. John Bradley (Florence Briggs Th...    Cherbourg
2    3rd                             Heikkinen, Miss. Laina  Southampton
3    1st       Futrelle, Mrs. Jacques Heath (Lily May Peel)  Southampton
4    3rd                           Allen, Mr. William Henry  Southampton

In [31]:
passengers.loc[10:20,['Age', 'Fare']] 
# passengers with indexes 10 to 20 passangers, only columns 'Age', 'Fare'

Age    Fare
10   4.0    16.7
11  58.0   26.55
12  20.0    8.05
13  39.0  31.275
14  14.0  7.8542
15  55.0    16.0
16   2.0  29.125
17  <NA>    13.0
18  31.0    18.0
19  <NA>   7.225
20  35.0    26.0

In [30]:
passengers.loc[:,['Name']] # All rows, column name

Name
0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
..                                                 ...
886                              Montvila, Rev. Juozas
887                       Graham, Miss. Margaret Edith
888           Johnston, Miss. Catherine Helen "Carrie"
889                              Behr, Mr. Karl Howell
890                                Dooley, Mr. Patrick

[891 rows x 1 columns]

In [32]:
# finally, you can combine them:
# get first 5 people, and their Names
passengers.iloc[0:5,:].loc[:,'Name']

0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: string

To pick first or last n items you can also use .head(n) or tail(n).

And if picking just one column (and it does not have spaces in the name), you can use .ColumnName 

In [33]:
passengers.head(5).Name

0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: string

In [34]:
# notice that .loc[:,'Name'] is the same as .Name
# try to guess, what are advantages and limitations of .ColumnName notation? 
passengers.head(5).loc[:,'Name']

0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: string

Ok, let;s find our 5 questions and answer them (call them A,B,C,D and E)

### A: Names of oldest passangers. What can we tell about their Survival rate when compared with youngest ones?

here's a start on the first one

In [35]:
# just select name and age components of the data frame. 
dfA2 = passengers[['Age', 'Name', 'Pclass', 'Survived']].copy()

# Sort them by Age 
dfA2.sort_values(by='Age', ascending=False, inplace=True)

# and get the names of 10 oldest passangers
oldest_10 = dfA2.iloc[0:10]
oldest_10

Age                                  Name Pclass  Survived
630  80.0  Barkworth, Mr. Algernon Henry Wilson    1st      True
851  74.0                   Svensson, Mr. Johan    3rd     False
493  71.0               Artagaveytia, Mr. Ramon    1st     False
96   71.0             Goldschmidt, Mr. George B    1st     False
116  70.5                  Connors, Mr. Patrick    3rd     False
672  70.0           Mitchell, Mr. Henry Michael    2nd     False
745  70.0          Crosby, Capt. Edward Gifford    1st     False
33   66.0                 Wheadon, Mr. Edward H    2nd     False
54   65.0        Ostby, Mr. Engelhart Cornelius    1st     False
280  65.0                      Duane, Mr. Frank    3rd     False

In [ ]:
# note: and if you wanted to get names of all people over 70
print(dfA2[dfA2.Age > 70].loc[:, ['Name','Age']])

In [36]:
# Or get names of all people over 50 who survived
print(dfA2[(dfA2.Age > 50) & (dfA2.Survived == 1)].loc[:, ['Name','Age']])

                                                  Name   Age
630               Barkworth, Mr. Algernon Henry Wilson  80.0
275                  Andrews, Miss. Kornelia Theodosia  63.0
483                             Turkula, Mrs. (Hedwig)  63.0
570                                 Harris, Mr. George  62.0
829          Stone, Mrs. George Nelson (Martha Evelyn)  62.0
587                   Frolicher-Stehli, Mr. Maxmillian  60.0
366   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  60.0
268      Graham, Mrs. William Thompson (Edith Junkins)  58.0
11                            Bonnell, Miss. Elizabeth  58.0
195                               Lurette, Miss. Elise  58.0
647                Simonius-Blumer, Col. Oberst Alfons  56.0
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  56.0
15                    Hewlett, Mrs. (Mary D Kingcome)   55.0
496                     Eustis, Miss. Elizabeth Mussey  54.0
774              Hocking, Mrs. Elizabeth (Eliza Needs)  54.0
513     Rothschild, Mrs.

In [45]:
print('oldest 10 survival', dfA2.iloc[0:10,:].loc[:, 'Survived'].mean())
# could also be
print('oldest 10 survival', dfA2.iloc[0:10,:].Survived.mean())
print('oldest 10 survival', dfA2.head(10).Survived.mean())

oldest 10 survival 0.1
oldest 10 survival 0.1
oldest 10 survival 0.1


In [47]:
print('youngest 10 survival', dfA2.tail(10).Survived.mean())

youngest 10 survival 0.3838383838383838


In [ ]:
# wait a minute.. that looks a bit suspicious! children died with such frequency?
# let's have a look at the data!
dfA2.tail(10)

In [48]:
# oh no! that's the data for those where age is missing!
# let's ask for average for those with know age, but still youngest
dfA2[ dfA2['Age'].notna()].tail(10)
# why do you think decimal ages are 0.42, 0.67, 0.75, 0.83 and 0.93 - oh why?

Age                             Name Pclass  Survived
381   1.0      Nakid, Miss. Maria ("Mary")    3rd      True
172   1.0     Johnson, Miss. Eleanor Ileen    3rd      True
164   1.0     Panula, Master. Eino Viljami    3rd     False
305  0.92   Allison, Master. Hudson Trevor    1st      True
831  0.83  Richards, Master. George Sibley    2nd      True
78   0.83    Caldwell, Master. Alden Gates    2nd      True
644  0.75           Baclini, Miss. Eugenie    3rd      True
469  0.75    Baclini, Miss. Helene Barbara    3rd      True
755  0.67        Hamalainen, Master. Viljo    2nd      True
803  0.42  Thomas, Master. Assad Alexander    3rd      True

In [49]:
print("youngest 10 survival: ", dfA2[ dfA2['Age'].notna()].tail(10).Survived.mean())

youngest 10 survival:  0.9


In [52]:
print('average survival rate', dfA2.Survived.mean())
print('average survival rate (if age known)', dfA2[ dfA2['Age'].notna()].Survived.mean())
# how do you interpret those results? What would it mean?

average survival rate 0.3838383838383838
average survival rate (if age known) 0.4061624649859944


Answer: Name of the oldest passanger (80) was Mr. Algernon Henry Wilson Barkworth. Survival rate of 10 oldest passangers was 10%, and of the youngest 10 passagners 90%. In context, the average survival rate was 38%.

### B: What's the impact of Passanger Class and Sex on the suvival?

In [ ]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### C: Your question? Select a subset, and look at it's qualities

In [ ]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### D: Your question?

In [ ]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### E: Your question?

In [ ]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### F: What is the mean and median age of passengers related to their survival status?

In [27]:
# Calculating the mean of the 'Age' column for each survival status
mean_age_survival=passengers.loc[:,['Survived','Age']].groupby('Survived').mean()
mean_age_survival

Age
Survived           
False     30.626179
True       28.34369

In [28]:
# Calculating the median of the 'Age' column for each survival status
median_age_survival=passengers.loc[:,['Survived','Age']].groupby('Survived').median()
median_age_survival

Age
Survived      
False     28.0
True      28.0

Our answer:
Mean age of passangers who survived was 28.34, and of those who did not 30.63. Median age of passangers who survived was 28.0, and of those who did not 28.0.
Median age is the same for both groups (28.0).

# Task 3: Cleaning up and improving the dataset:

A: Identify missing values and decide what to do with them. You can remove whole rows with missing things, or you can fill them with averages. Or averages for a particular subgroup.

B: Rename the column names so that they follow the same naming convention (eg. no spaces, CamelCase or snake_case). 

C: Can you create new columns which use available information? eg. Can you create a column TravelingAlone based on the values in siblings/partners and parents/children ?

### A: Looking at missing values

In [ ]:
# btw. re-run this cell to reset yoru dataset it its original form
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')


In [82]:
# all data into binary: is missing or not:
passengers.isna()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Embarked, Length: 889, dtype: bool

In [90]:
# by columns, are any items True?
passengers.isna().any()

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [59]:
# how many passagners would you loose if you passengers.dropna(axis=0) ?
# you can use .shape()
pass_no_na = passengers.dropna(axis=0)
print("Dropped rows:", len(passengers) - len(pass_no_na))

Dropped rows: 708


In [63]:
# Calculating the mean of the 'Age' column for each survival status
pass_no_na_mean= pass_no_na.loc[:, ['Survived', 'Age']].groupby('Survived').mean()
print(pass_no_na_mean)

# Calculating the median of the 'Age' column for each survival status
pass_no_na_median=pass_no_na.loc[:,['Survived','Age']].groupby('Survived').median()
print(pass_no_na_median)

                Age
Survived           
False         41.35
True      32.905854
           Age
Survived      
False     44.5
True      33.0


In [64]:
# You could replace all missing values with a set value, 
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
pass_no_na = passengers.fillna(0) 
# but what unexpected effects it could have?
pass_no_na

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   0.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500     0        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250     0        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500     0        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000     0        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500     0        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500     0        Q  

[891 rows x 12 columns]

In [65]:
# or you can replace them with a most common value,
# or other statistic of that column eg.mean, max
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
passengers.fillna(passengers.mean(numeric_only=True)) 

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex        Age  \
0                              Braund, Mr. Owen Harris    male  22.000000   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.000000   
2                               Heikkinen, Miss. Laina  female  26.000000   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.000000   
4                             Allen, Mr. William Henry    male  35.000000   
..                                                 ...     ...        ...   
886                              Montvila, Rev. Juozas    male  27.000000   
887                       Graham, Miss. Margaret Edith  female  19.000000   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.699118   
889                              Behr, Mr. Karl Howell    male  26.000000   
890                                Dooley, Mr. Patrick    male  32.000000   

     SibSp  Parch            Ticket     Fare Cabin Embarked  
0        1      0         A/5 21171   7.2500   NaN        S  
1        1      0          PC 17599  71.2833   C85        C  
2        0      0  STON/O2. 3101282   7.9250   NaN        S  
3        1      0            113803  53.1000  C123        S  
4        0      0            373450   8.0500   NaN        S  
..     ...    ...               ...      ...   ...      ...  
886      0      0            211536  13.0000   NaN        S  
887      0      0            112053  30.0000   B42        S  
888      1      2        W./C. 6607  23.4500   NaN        S  
889      0      0            111369  30.0000  C148        C  
890      0      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [85]:
# for non-numeric values one of the options is to use the most common (mode) value.
# since there might be mady most common values, you'll have to specify that you want the first one
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
passengers.fillna(passengers.mode().iloc[0]) 
# notice what happens with the Cabin

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  24.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare    Cabin Embarked  
0        0         A/5 21171   7.2500  B96 B98        S  
1        0          PC 17599  71.2833      C85        C  
2        0  STON/O2. 3101282   7.9250  B96 B98        S  
3        0            113803  53.1000     C123        S  
4        0            373450   8.0500  B96 B98        S  
..     ...               ...      ...      ...      ...  
886      0            211536  13.0000  B96 B98        S  
887      0            112053  30.0000      B42        S  
888      2        W./C. 6607  23.4500  B96 B98        S  
889      0            111369  30.0000     C148        C  
890      0            370376   7.7500  B96 B98        Q  

[891 rows x 12 columns]

In [86]:
# you could also sopecify individual columns and values
#passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
passengers.Cabin = passengers.Cabin.fillna("NoCabin") 
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare    Cabin Embarked  
0        0         A/5 21171   7.2500  NoCabin        S  
1        0          PC 17599  71.2833      C85        C  
2        0  STON/O2. 3101282   7.9250  NoCabin        S  
3        0            113803  53.1000     C123        S  
4        0            373450   8.0500  NoCabin        S  
..     ...               ...      ...      ...      ...  
886      0            211536  13.0000  NoCabin        S  
887      0            112053  30.0000      B42        S  
888      2        W./C. 6607  23.4500  NoCabin        S  
889      0            111369  30.0000     C148        C  
890      0            370376   7.7500  NoCabin        Q  

[891 rows x 12 columns]

In [ ]:
# Your task, decide what to do with the missing values in the origibnal dataset:
# Once your_dataset.isna().any() returns all False, there are no more missing values

In [89]:
passengers.Age = passengers.Age.fillna(np.ceil(passengers.Age.mean()))
passengers = passengers.dropna(axis=0)
print(passengers.shape)
print(passengers.isna().any())

(889, 12)
PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age            False
SibSp          False
Parch          False
Ticket         False
Fare           False
Cabin          False
Embarked       False
dtype: bool


### B: Duplicates. 

In this dataset there are no dulplicates. So here's a small exercise. Let's duplicate the first person so they appear 3 times. And let's change the age of one of those entries. So we're pretending first passanger got accidentally added 3 times, including once with a slightly different data for age.

In [90]:
# att two more copies of first row
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
with_duplicates = passengers
with_duplicates = pd.concat((passengers.iloc[0:1], passengers.iloc[0:1], passengers))
with_duplicates

PassengerId  Survived  Pclass  \
0              1         0       3   
0              1         0       3   
0              1         0       3   
1              2         1       1   
2              3         1       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
0                              Braund, Mr. Owen Harris    male  22.0      1   
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
0        0         A/5 21171   7.2500   NaN        S  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[893 rows x 12 columns]

In [91]:
# now let's fix the index
with_duplicates.reset_index(drop=True, inplace=True) # drop completely forgets previous index
with_duplicates

PassengerId  Survived  Pclass  \
0              1         0       3   
1              1         0       3   
2              1         0       3   
3              2         1       1   
4              3         1       3   
..           ...       ...     ...   
888          887         0       2   
889          888         1       1   
890          889         0       3   
891          890         1       1   
892          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1                              Braund, Mr. Owen Harris    male  22.0      1   
2                              Braund, Mr. Owen Harris    male  22.0      1   
3    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                               Heikkinen, Miss. Laina  female  26.0      0   
..                                                 ...     ...   ...    ...   
888                              Montvila, Rev. Juozas    male  27.0      0   
889                       Graham, Miss. Margaret Edith  female  19.0      0   
890           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
891                              Behr, Mr. Karl Howell    male  26.0      0   
892                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0         A/5 21171   7.2500   NaN        S  
2        0         A/5 21171   7.2500   NaN        S  
3        0          PC 17599  71.2833   C85        C  
4        0  STON/O2. 3101282   7.9250   NaN        S  
..     ...               ...      ...   ...      ...  
888      0            211536  13.0000   NaN        S  
889      0            112053  30.0000   B42        S  
890      2        W./C. 6607  23.4500   NaN        S  
891      0            111369  30.0000  C148        C  
892      0            370376   7.7500   NaN        Q  

[893 rows x 12 columns]

In [94]:
# change one of the duplicates to have another age:
with_duplicates.loc[1,'Age'] = 23 

# Note: as well as loc[] and iloc[], which access many items,
# you can use at[] and iat[] to access just one:
with_duplicates.at[1,'Age'] = 24
with_duplicates

PassengerId  Survived  Pclass  \
0              1         0       3   
1              1         0       3   
2              1         0       3   
3              2         1       1   
4              3         1       3   
..           ...       ...     ...   
888          887         0       2   
889          888         1       1   
890          889         0       3   
891          890         1       1   
892          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1                              Braund, Mr. Owen Harris    male  24.0      1   
2                              Braund, Mr. Owen Harris    male  22.0      1   
3    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                               Heikkinen, Miss. Laina  female  26.0      0   
..                                                 ...     ...   ...    ...   
888                              Montvila, Rev. Juozas    male  27.0      0   
889                       Graham, Miss. Margaret Edith  female  19.0      0   
890           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
891                              Behr, Mr. Karl Howell    male  26.0      0   
892                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0         A/5 21171   7.2500   NaN        S  
2        0         A/5 21171   7.2500   NaN        S  
3        0          PC 17599  71.2833   C85        C  
4        0  STON/O2. 3101282   7.9250   NaN        S  
..     ...               ...      ...   ...      ...  
888      0            211536  13.0000   NaN        S  
889      0            112053  30.0000   B42        S  
890      2        W./C. 6607  23.4500   NaN        S  
891      0            111369  30.0000  C148        C  
892      0            370376   7.7500   NaN        Q  

[893 rows x 12 columns]

Now let's check for duplicates, and remove them. You might want to define a duplicate as when some unique element (or a combination of unique elements) are the same.

In [95]:
with_duplicates.duplicated()
# here you should see that the person at index 2 is a duplicate (perfectly identical)
# you could also decide to keep last (not first) with  duplicated(keep='last')

0      False
1      False
2       True
3      False
4      False
       ...  
888    False
889    False
890    False
891    False
892    False
Length: 893, dtype: bool

In [97]:
no_duplicates1 = with_duplicates.drop_duplicates()
no_duplicates1
# this would remove all completely identical duplicates

PassengerId  Survived  Pclass  \
0              1         0       3   
1              1         0       3   
3              2         1       1   
4              3         1       3   
5              4         1       1   
..           ...       ...     ...   
888          887         0       2   
889          888         1       1   
890          889         0       3   
891          890         1       1   
892          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1                              Braund, Mr. Owen Harris    male  24.0      1   
3    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                               Heikkinen, Miss. Laina  female  26.0      0   
5         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
..                                                 ...     ...   ...    ...   
888                              Montvila, Rev. Juozas    male  27.0      0   
889                       Graham, Miss. Margaret Edith  female  19.0      0   
890           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
891                              Behr, Mr. Karl Howell    male  26.0      0   
892                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0         A/5 21171   7.2500   NaN        S  
3        0          PC 17599  71.2833   C85        C  
4        0  STON/O2. 3101282   7.9250   NaN        S  
5        0            113803  53.1000  C123        S  
..     ...               ...      ...   ...      ...  
888      0            211536  13.0000   NaN        S  
889      0            112053  30.0000   B42        S  
890      2        W./C. 6607  23.4500   NaN        S  
891      0            111369  30.0000  C148        C  
892      0            370376   7.7500   NaN        Q  

[892 rows x 12 columns]

In [96]:
# this would remove all duplicates, identical only in those columns
no_duplicates2 = with_duplicates.drop_duplicates(subset=['Name'])
no_duplicates2
# notice we're down to one duplicated first passanger

PassengerId  Survived  Pclass  \
0              1         0       3   
3              2         1       1   
4              3         1       3   
5              4         1       1   
6              5         0       3   
..           ...       ...     ...   
888          887         0       2   
889          888         1       1   
890          889         0       3   
891          890         1       1   
892          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
3    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                               Heikkinen, Miss. Laina  female  26.0      0   
5         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
888                              Montvila, Rev. Juozas    male  27.0      0   
889                       Graham, Miss. Margaret Edith  female  19.0      0   
890           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
891                              Behr, Mr. Karl Howell    male  26.0      0   
892                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
3        0          PC 17599  71.2833   C85        C  
4        0  STON/O2. 3101282   7.9250   NaN        S  
5        0            113803  53.1000  C123        S  
6        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
888      0            211536  13.0000   NaN        S  
889      0            112053  30.0000   B42        S  
890      2        W./C. 6607  23.4500   NaN        S  
891      0            111369  30.0000  C148        C  
892      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [98]:
# so now there are no duplicates
no_duplicates2.duplicated()

0      False
3      False
4      False
5      False
6      False
       ...  
888    False
889    False
890    False
891    False
892    False
Length: 891, dtype: bool

In [99]:
# and we can re-index
no_duplicates2.reset_index(drop=True, inplace=True)
no_duplicates2

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### C: Improving Column names

Look at the notes if you're unsure how to do it. You're looking for rename(). Aim for no spaces or special characters in column names, so that you can use . notation, like .Age

One improvementc could be making those columns more human readable: SibSp, Parch. See link to Kaggle above for more insights or what they mean/

In [100]:
passengers = passengers.rename(columns = {'Pclass': 'Passenger_Class', 'SibSp': 'Siblings_Spouses', 'Parch': 'Parents_Children'})
passengers

PassengerId  Survived  Passenger_Class  \
0              1         0                3   
1              2         1                1   
2              3         1                3   
3              4         1                1   
4              5         0                3   
..           ...       ...              ...   
886          887         0                2   
887          888         1                1   
888          889         0                3   
889          890         1                1   
890          891         0                3   

                                                  Name     Sex   Age  \
0                              Braund, Mr. Owen Harris    male  22.0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
2                               Heikkinen, Miss. Laina  female  26.0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
4                             Allen, Mr. William Henry    male  35.0   
..                                                 ...     ...   ...   
886                              Montvila, Rev. Juozas    male  27.0   
887                       Graham, Miss. Margaret Edith  female  19.0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
889                              Behr, Mr. Karl Howell    male  26.0   
890                                Dooley, Mr. Patrick    male  32.0   

     Siblings_Spouses  Parents_Children            Ticket     Fare Cabin  \
0                   1                 0         A/5 21171   7.2500   NaN   
1                   1                 0          PC 17599  71.2833   C85   
2                   0                 0  STON/O2. 3101282   7.9250   NaN   
3                   1                 0            113803  53.1000  C123   
4                   0                 0            373450   8.0500   NaN   
..                ...               ...               ...      ...   ...   
886                 0                 0            211536  13.0000   NaN   
887                 0                 0            112053  30.0000   B42   
888                 1                 2        W./C. 6607  23.4500   NaN   
889                 0                 0            111369  30.0000  C148   
890                 0                 0            370376   7.7500   NaN   

    Embarked  
0          S  
1          C  
2          S  
3          S  
4          S  
..       ...  
886        S  
887        S  
888        S  
889        C  
890        Q  

[891 rows x 12 columns]

### D: Categorical variables (in R called Vectors, in Java called Enums)

Some values are presented as numbers (Pclass), but really, they are categories. Some are presented as strings (sex), but with some careful consideration and kind design can be analysed as categories too (notice this very old dataset is using Sex, not gender*).

**Note on usage of Gender and Sex in datsets: read up on this after the lab**

*In all modern datasets Gender is represented as much more complex datastructure than just True/False binary (this was common in early days of databases, where gender was a True/False value 'is_male'). Later it was a category (with 2 or 3 choices eg. 'male', 'female', 'prefer_not_to_say'). 

Luckilly, to reflect beauty of human experience there are some very interesting research projects on how to design inclusive forms and datsets. This Scottish Government oriject involved data scientists and people with various experiences of gender. Do some reading on the subject https://www.gov.scot/publications/reviewing-design-methods-make-more-sensitive-gender/pages/1/ and notice it does not provide 'one answer fits all' but rather encourages empathy on project-by-project basis. Also form designers together with LGBTQI+ organisations are constanly advancing this field https://uxdesign.cc/designing-forms-for-gender-diversity-and-inclusion-d8194cf1f51  

So let's not look at the Sex field, but rather Pclass.

Let's first have a look at exploring Passanger class: What passanger classess were there on the titanic (on british trains we still have 1st and 2nd class carriages

In the dataset they are represented 

In [109]:
# Loading the dataset again
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')

In [110]:
# Showing the Pclass column
passengers.Pclass

0      3
1      1
2      3
3      1
4      3
      ..
886    2
887    1
888    3
889    1
890    3
Name: Pclass, Length: 891, dtype: int64

In [111]:
passengers.Pclass.value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [112]:
# are there any NA ?
passengers.Pclass.isna().value_counts()

Pclass
False    891
Name: count, dtype: int64

In [113]:
passengers['TicketClass'] = pd.Categorical(passengers.Pclass)
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked TicketClass  
0        0         A/5 21171   7.2500   NaN        S           3  
1        0          PC 17599  71.2833   C85        C           1  
2        0  STON/O2. 3101282   7.9250   NaN        S           3  
3        0            113803  53.1000  C123        S           1  
4        0            373450   8.0500   NaN        S           3  
..     ...               ...      ...   ...      ...         ...  
886      0            211536  13.0000   NaN        S           2  
887      0            112053  30.0000   B42        S           1  
888      2        W./C. 6607  23.4500   NaN        S           3  
889      0            111369  30.0000  C148        C           1  
890      0            370376   7.7500   NaN        Q           3  

[891 rows x 13 columns]

In [114]:
# Categorical data has number of advantages in generating statistics etc
# but also make our datasets much smaller (in memory) and faster (in processing)
passengers.memory_usage()
# notice ticketclass stores the same data as Pclass, but uses 15% of memory.

Index           132
PassengerId    7128
Survived       7128
Pclass         7128
Name           7128
Sex            7128
Age            7128
SibSp          7128
Parch          7128
Ticket         7128
Fare           7128
Cabin          7128
Embarked       7128
TicketClass    1023
dtype: int64

In [116]:
# Additionally you can specify the order of categories, so that you 
# could perform comparisons

# introduce order
passengers['TicketClass'] = pd.Categorical(passengers.Pclass, 
                                           categories=[3,2,1], # here specify order
                                           ordered=True
                                          )
# optionally re-name categories with cat.rename_categories
passengers['TicketClass'] = passengers['TicketClass'].cat.rename_categories(
  {1: 'Luxury', 2: 'Economy', 3: 'LowerDeck'})
passengers['TicketClass'] # notice printout shows order of categories

0      LowerDeck
1         Luxury
2      LowerDeck
3         Luxury
4      LowerDeck
         ...    
886      Economy
887       Luxury
888    LowerDeck
889       Luxury
890    LowerDeck
Name: TicketClass, Length: 891, dtype: category
Categories (3, object): ['LowerDeck' < 'Economy' < 'Luxury']

In [118]:
# so that you can now eg choose all TicketClass LOWER THAN a number:
passengers[passengers['TicketClass'] < 'Luxury']

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
5      male   NaN      0      0            330877   8.4583   NaN        Q   
7      male   2.0      3      1            349909  21.0750   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S   
885  female  39.0      0      5            382652  29.1250   NaN        Q   
886    male  27.0      0      0            211536  13.0000   NaN        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

    TicketClass  
0     LowerDeck  
2     LowerDeck  
4     LowerDeck  
5     LowerDeck  
7     LowerDeck  
..          ...  
884   LowerDeck  
885   LowerDeck  
886     Economy  
888   LowerDeck  
890   LowerDeck  

[675 rows x 13 columns]

### E: Creating more columns

Look at the notes if you're unsure how to do it. What could you use: age groups? In one of the notebooks you were shown how to group different ranges of values into groups. Maybe you could use that technique on age?

Or maybe you could combine columns abotu partners and dependants (SibSp, Parch) into a column TravelsAlone?

How do Survival stats look like for different values of the new columns?